### PR TITLE
feat: update the metamask sdk package to version "0.18.3"

### PR DESCRIPTION
**Pull request by omridan159**
## Description

This PR updates the `MetaMask SDK` package to version `0.18.3`, introducing significant improvements aimed at reducing the bundle size and enhancing the MetaMask connector's logic. This version marks a pivotal update where we've focused on optimizing the overall package size to ensure faster loading times and a more efficient integration process for developers. Alongside the SDK bundle size reduction, this PR addresses various bugs and enhances the SDK's functionality to operate seamlessly without the necessity for the `extensionOnly=true` option.

It is also crucial to highlight that `dappMetadata` has become mandatory for integration with this version. This change underscores our commitment to improving security and user experience by ensuring that dApps provide the necessary information for a safer and more transparent interaction.

Furthermore, to expand the SDK's compatibility and usability on mobile devices, we have introduced support for custom RPC URLs. This enhancement allows for the direct incorporation of custom RPC URLs from the Wagmi Config, offering developers the flexibility to specify their network configurations, thereby facilitating a smoother integration with different blockchain networks.

These updates collectively contribute to a more robust, efficient, and versatile MetaMask SDK, setting a new standard for blockchain application development.

## Additional Information

Before submitting this issue, please make sure you do the following.

- [x] Read the [contributing guide](https://wagmi.sh/dev/contributing)
- [x] Added documentation related to the changes made.
- [x] Added or updated tests (and snapshots) related to the changes made.

### CODE REVIEW

File: packages/connectors/src/metaMask.ts [hunk 1]
```diff
@@ -51,47 +47,40 @@ export function metaMask(parameters: MetaMaskParameters = {}) {
...
+        sdkImport = import('@metamask/sdk') as any
+      }
+
+      const sdkModule = await sdkImport
+      if (!sdkModule) {
+        throw new Error('MetaMask SDK not found')
+      }
+
+      sdk = new sdkModule.MetaMaskSDK({
+        dappMetadata: { name: 'wagmi' }, // Must be set and overwritten via parameters
```
Comment (abretonc7s) [inline packages/connectors/src/metaMask.ts:66]: please remove this line, user must set it in wagmi

**Review by abretonc7s — CHANGES_REQUESTED** [review]
small change for dapp metadata

**Review by abretonc7s — APPROVED** [review]
LGTM

File: packages/create-wagmi/templates/next/src/wagmi.ts [hunk 2]
```diff
@@ -1,11 +1,22 @@
...
 
 export const config = createConfig({
   chains: [mainnet, sepolia],
   connectors: [
     injected(),
+    metaMask({
+      dappMetadata: {
+        name: 'Create Wagmi',
+      },
+      infuraAPIKey: import.meta.env.VITE_INFURA_API_KEY,
```
Comment (jxom) [inline packages/create-wagmi/templates/next/src/wagmi.ts:18]: I don't think we should expose this at all as not all consumers use Infura, and tend to bring their own RPC providers. Instead, can we have the ability to pass an arbitrary RPC URL to the MetaMask SDK? This is what we are doing for the other connectors, like WalletConnect: https://github.com/wevm/wagmi/blob/main/packages/connectors/src/walletConnect.ts#L212-L217

This means we can infer the RPC URL from the provided `transports` from the Wagmi Config.

File: packages/connectors/src/metaMask.ts [hunk 3]
```diff
@@ -35,14 +34,11 @@ export type MetaMaskParameters = Evaluate<
   >
 >
 
+let sdk: MetaMaskSDK
+let sdkImport: Promise<typeof import('@metamask/sdk')> | undefined
```
[the remark this answers is not in the document]
Comment (deeeed) [reply packages/connectors/src/metaMask.ts:38]: yes we can move it back, was initially testing different variation but eventually the scope shouldn't have impact.

File: packages/create-wagmi/templates/next/src/wagmi.ts (hunk 2 above)
Comment (deeeed) [reply packages/create-wagmi/templates/next/src/wagmi.ts:18]: Thanks, we have a similar way to configure rpcMap, will update accordingly.

**Comment by jxom** [comment]
Has there been any efforts so far to mitigate bundle size? I am seeing no difference between `0.14.3` and `0.15.0`:

### 0.15.0

```
  Size:         324.98 kB with all dependencies, minified and brotlied
  Loading time: 6.4 s     on slow 3G
  Running time: 7.1 s     on Snapdragon 410
  Total time:   13.5 s
 ```
 
 ### 0.14.3

```
  Size:         324.61 kB with all dependencies, minified and brotlied
  Loading time: 6.4 s     on slow 3G
  Running time: 1.8 s     on Snapdragon 410
  Total time:   8.2 s
 ```

**Comment by deeeed** [comment]
> Will review this in-depth later in the week and happy to merge after that.
> 
> I was also wondering – has there been any efforts so far to mitigate bundle size? Saw that in the SDK changelog, but I am seeing no difference between `0.14.3` and `0.15.0` (looks like it increased very slightly?):
> 
> ### 0.15.0
> ```
>   import { MetaMaskSDK } from '@metamask/sdk'
>   //       ^?
> 
>   Size:         324.98 kB with all dependencies, minified and brotlied
>   Loading time: 6.4 s     on slow 3G
>   Running time: 7.1 s     on Snapdragon 410
>   Total time:   13.5 s
> ```
> 
> ### 0.14.3
> ```
>   import { MetaMaskSDK } from '@metamask/sdk'
>   //       ^?
>   
>   Size:         324.61 kB with all dependencies, minified and brotlied
>   Loading time: 6.4 s     on slow 3G
>   Running time: 1.8 s     on Snapdragon 410
>   Total time:   8.2 s
> ```
> 
> Reproduced with [`size-limit`](https://github.com/ai/size-limit)

Not on the source itself, we are tackling that next. Currently we removed the sourcemaps that were creating high package size on npm and was potentially scaring people away (even though it is unrelated). We will have another update on the PR today to address your comments.

**Pull request closed without merging**